Standardise and fix parsing and serialization of blocks and nav menu items

getdave opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
Aims to fix https://github.com/WordPress/gutenberg/issues/29793.

This PR does two things:

1. Updates the Nav Editor _and_ Nav Block to create the correct `core/navigation-link` block variations based on the menu item data from the database.
2. Correctly converts the blocks back to matching nav menu objects for storage in the database, preserving the data necessary to ensure they:
  - can be converted back into the correct `core/navigation-link` block _variation_.
  - have the correct object id assigned to match the entity ID which they represent (eg: a `Post` or a `Page`...etc).

### Known Bugs

**Note** in the process of working on this PR I've uncovered a few more bugs which I won't attempt to fix in this PR lest it get too large. I'm noting for future reference here:
* Post type variations display as `Post link
` in the block sidebar ([screenshot](https://user-images.githubusercontent.com/444434/115868981-266f9180-a435-11eb-8707-ac75528d7ad2.png)). They should probably use the title of the Custom Post Type (eg: `Portfolio link`)
* There is no post type variation for `Post Type Archive` as a result these are saved as `Custom`. We need a new variation to be created. [Here's how you add a `Post Type Archive` to a menu](https://www.wpbeginner.com/wp-tutorials/how-to-add-post-type-archive-in-wordpress-navigation-menus/). See https://github.com/WordPress/gutenberg/issues/31452
* ~Other attributes aren't always parsed/serialized correctly from block -> menuItem and menuItem -> block (eg: `attr_label`...etc). Once this PR is merged, we need to audit these and write some automated tests to ensure they are adding to the mapping for conversion.~ **Update**: this is now resolved as part of this PR.
* Post Format's save as `Custom` because the REST API endpoint for `v2/search` doesn't return numeric IDs for the post formats. See https://github.com/WordPress/gutenberg/issues/31380

## How has this been tested?

### Setup
* Install this Plugin on your local env https://wordpress.org/plugins/portfolio-post-type/ and activate.
* Create some Posts, Pages and Categories - I used FakerPress plugin.
* Create some Portfolio items, Portfolio Categories and Porfolio Tags - I used FakerPress plugin.
* Create a menu at http://localhost:8889/wp-admin/nav-menus.php and populate with a mixture of:
  - Posts
  - Pages
  - Categories
  - Tags
  - Custom Links
  - Portfolio posts.
  - Portfolio Categories
  - Portfolio Archive ([here's how you do that](https://www.wpbeginner.com/wp-tutorials/how-to-add-post-type-archive-in-wordpress-navigation-menus/))
* Save your Menu!

### Test
* Open the Navigation editor at http://localhost:8889/wp-admin/admin.php?page=gutenberg-navigation.
* You should see the nav menu appear and all the `core/navigation-link`s should appear with the correct block variation (eg: `Post Link, `Page Link`, `Custom`...etc. There are a few known exceptions that don't work which - given that they have _never_ worked - I will tackle in a follow up. These are:
  - `Post Type Archive` this will not work because we don't have a variation setup for this yet.
  - Post Type Categories (eg: anything you created under `Portfolio Categories` will show up with the block variation name being `Category`. This needs fixing.
* Save your Navigation - this will resave all the items that are currently blocks in the Navigation Editor back to menu items in the database. This will expose any saving errors.
* Go back to the old Menus page at http://localhost:8889/wp-admin/nav-menus.php.
* Check the menu items have been correctly persisted and retained their original "types". 


## Screenshots <!-- if applicable -->

### Menu -> Nav Editor -> Menu
https://user-images.githubusercontent.com/444434/115868532-8154b900-a434-11eb-9c04-03846ff0fd91.mov

### Menu -> Nav Block

https://user-images.githubusercontent.com/444434/115868720-c973db80-a434-11eb-9f43-71a97716d5d6.mov





## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

getdave committed:
Create correct nav link block variations from menu data type

getdave committed:
Set menu object type based on block type

getdave committed:
Correctly hydrate nav block itself from stored menu object fields.

getdave commented on packages/block-library/src/navigation-link/hooks.js:
```diff
@@ -46,7 +46,11 @@ export function enhanceNavigationLinkVariations( settings, name ) {
 	if ( settings.variations ) {
 		const isActive = ( blockAttributes, variationAttributes ) => {
-			return blockAttributes.type === variationAttributes.type;
+			const normalizedVariationKind = variationAttributes.kind.replace(
```
I'm not clear why we have two versions of this code...

talldan replied:
I think the fallback variations will be removed in a later WP version, they're being kept for back compat.

gwwar replied:
The fallback is only used for versions of WP 5.7 and below. Using WP 5.8+ variations are provided by navigation-link in `register_block_type_from_metadata`

https://github.com/WordPress/gutenberg/blob/e65038afdf575eede1b6ebde37afd18e977b1c0d/packages/block-library/src/navigation-link/index.php#L293-L333

We'll be able to remove the entire hook.js file, after [`isActive` supports an array of strings](https://github.com/WordPress/gutenberg/pull/30913) as option, icons map to a string, AND the when the gutenberg plugin moves past <= 5.7 support.

talldan commented on packages/block-library/src/navigation-link/fallback-variations.js:
```diff
@@ -58,8 +58,14 @@ const fallbackVariations = [
-	variation.isActive = ( blockAttributes, variationAttributes ) =>
-		blockAttributes.type === variationAttributes.type;
+
+	variation.isActive = ( blockAttributes, variationAttributes ) => {
+		const normalizedVariationKind = variationAttributes.kind.replace(
+			'-',
+			'_'
+		);
+		return blockAttributes.type === normalizedVariationKind;
+	};
```
I'm not sure comparing the kind to the type like this will work (e.g. for a post the kind is `postType` and the type `post`).

What was the problem that required this change?

getdave replied:
Thanks for calling this out. Turns out I needed to be storing/retrieveing data from the `object` field instead. With that change we can now revert both the variations `isActive` comparisons to how they were previously.

getdave committed:
Fix to store/retrieve nav link type from/to the menu item’s “object” field.

This field is the correct field in which to store the type of the object being stored (eg: “post”, “pages”…etc).

See https://core.trac.wordpress.org/browser/tags/5.7.1/src/wp-includes/nav-menu.php#L796

getdave committed:
Fix up Nav Block to use object field for type

getdave committed:
Avoid storing menu objects with “custom” as type.

getdave committed:
Only set id and type on the generated block if the object and object_id fields are provided in the menu object

This avoids “undefined” values.

getdave committed:
Add test to cover correct creation of core/navigation-link block variations from menu objects

getdave committed:
Conditional persist block id and type attrs to menu object fields

getdave committed:
Add tests to cover conversion from blocks to menu objects

getdave committed:
Revert changes to fallback variations isActive detection logic

getdave committed:
Remove accidental new line in file causes “changes”

getdave commented:
I've realised this is more complex than I thought.

Essentially we have x2 fields on the `nav_menu_item` coming from the database

1. `type` - can either be: `post_type`, `post_type_archive`, `taxonomy` (otherwise it falls back to `custom`).
2. `object`  - refers to the _specific_ object type eg: `post`, `page`.

We also have `object_id` which is a reference to the ID of the `object`. This is currently mapped to the `block.attributes.id` field.

On our `core/navigation-link` block we have:

1. `type` - this PR updates it to map to the `object` field.
2. `kind` - this sort of maps to the `type` field on the `nav_menu_item` but the format is different.

I think we need to update everything as follows:

* Update `type` to map 1:1 from block to `nav_menu_item` object.
* Add a new `object` attribute to the `block.attirbutes` that maps 1:1 with the `object` field on the `nav_menu_item`.
* Update the variations so that the variation is matched against the `object` type which is the most specific identifier to the "type" of the item.

I'll dive into this again tomorrow.

getdave commented:
Watch out for 

1. post formats. testing instructoins in https://github.com/WordPress/gutenberg/pull/29692
2. post type archives - https://www.wpbeginner.com/wp-tutorials/how-to-add-post-type-archive-in-wordpress-navigation-menus/

gwwar commented:
> On our core/navigation-link block we have:

Leaving a brain dump of the current values. I think all props are optional (we render placeholders in the editor). Some older data might be missing `kind` information.

`label` : This the text value of the link. May contain rich text formatting.
`type` :  `{ 'post' | 'page' | 'category' | 'tag' }` or `$taxonomy->name` or `$post_type->name` for a CPT. Note the exception for `tag` instead of `post_tag`.
`description`: Menu description if the theme supports it
`rel`: link rel, see https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types
`id`: term id OR post id. Also an optional attribute.
`opensInNewTab`: sets link `target="_blank"`
`url`: link url is set even if we have enough information to generate this, for example having an id, type and kind value. This is probably open for iteration. (A custom link is simply a label+url set, but we can be more explicit in noting the custom type in `kind`.)
`title`: the [title attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/title) set on a link
`kind`: `{ 'post-type' | 'taxonomy' }` `post-format` is also passed around in the kind variable internally, but doesn't get serialized. We're likely missing `post_type_archive` support.

https://github.com/WordPress/gutenberg/blob/ee84e9447f3f148684d29573558a69314b90274e/packages/block-library/src/navigation-link/block.json#L1-L35

gwwar commented:
I think we have some flexibility in the data formats, what might help drive this is adding unit test cases for our common scenarios. If it makes sense, I think we'll want three suites:

- MenuData -> MenuData ( Are values retained when I save in the navigation editor? This is especially noticeable when going from customizer/old menu editor -> navigation editor  )
- MenuData-> NavigationLink (When I load a menu in the site/post editor using the navigation block).
- NavigationLink -> MenuData (A user started from a named menu in the site/post editor, we then edit the menu in the navigation editor).

Each suite would likely cover cases like:
- Custom Links
- PostType/Taxonomy Links
- Tag test for the hardcoded value
- Drafts
- Home Link (Might need to wait for this to be fully implemented/a bit further along).

getdave committed:
Create mapping utils to map menu items to blocks and vice versa

getdave committed:
Genericise conversion util and share

getdave committed:
Doc block for conversion util

getdave committed:
Add simple conversion of kind on the nav block itself

getdave committed:
Correct type to kind conversion on block

getdave committed:
Fix up

getdave committed:
Simplify mapper function

getdave committed:
Fix broken test

getdave committed:
Extract mapping functions to shared utils and test

getdave commented:
Currently working on fixing the e2e tests.

getdave committed:
Fix snapshot test to account for new attributes

getdave commented:
Still working through the e2e tests and fixing issue with `id` attribute being persisted for links with `custom` type.

getdave committed:
Only map id if it is not a custom link

getdave committed:
Update nav block snapshot to include new attributes

getdave committed:
Fix test that could never have passed

getdave committed:
Allow for pass through mapping if a mapping is not defined for this attribute

getdave commented on packages/block-library/src/navigation/placeholder.js:
```diff
@@ -56,6 +56,12 @@ function mapMenuItemsToBlocks( menuItems ) {
 				: menuItem.title.rendered,
 			opensInNewTab: menuItem.target === '_blank',
+			type: menuItem?.object,
+			...( menuItem?.object !== 'custom' && {
```
Only set id if it's an entity type (ie: not the `custom` type).

gwwar replied:
Minor: It looks like we're trying to avoid adding an id prop. When menuItem.object is undefined this will set `id` to undefined. Eg `{ id:undefined }` instead of `{}`

getdave replied:
Fixed as part of refactor.

getdave commented on packages/e2e-tests/specs/experiments/__snapshots__/navigation-editor.test.js.snap:
```diff
@@ -10,36 +10,36 @@ exports[`Navigation editor allows creation of a menu when there are no current m
-<!-- wp:navigation-link {\\"label\\":\\"Fuga odio quis tempora\\",\\"description\\":\\"\\",\\"rel\\":\\"\\",\\"url\\":\\"http://localhost:8889/?page_id=56\\",\\"title\\":\\"\\",\\"className\\":\\"\\"} -->
-<!-- wp:navigation-link {\\"label\\":\\"In consectetur repellendus eveniet maiores aperiam\\",\\"description\\":\\"\\",\\"rel\\":\\"\\",\\"url\\":\\"http://localhost:8889/?page_id=15\\",\\"title\\":\\"\\",\\"className\\":\\"\\"} -->
-<!-- wp:navigation-link {\\"label\\":\\"Mollitia maiores consequatur ea dolorem blanditiis\\",\\"description\\":\\"\\",\\"rel\\":\\"\\",\\"url\\":\\"http://localhost:8889/?page_id=45\\",\\"title\\":\\"\\",\\"className\\":\\"\\"} -->
-<!-- wp:navigation-link {\\"label\\":\\"Necessitatibus nisi qui qui necessitatibus quaerat possimus\\",\\"description\\":\\"\\",\\"rel\\":\\"\\",\\"url\\":\\"http://localhost:8889/?page_id=27\\",\\"title\\":\\"\\",\\"className\\":\\"\\"} /-->
+<!-- wp:navigation-link {\\"label\\":\\"Est ea vero non nihil officiis in\\",\\"type\\":\\"page\\",\\"description\\":\\"\\",\\"rel\\":\\"\\",\\"id\\":53,\\"url\\":\\"http://localhost:8889/?page_id=53\\",\\"title\\":\\"\\",\\"kind\\":\\"post-type\\",\\"className\\":\\"\\"} -->
+<!-- wp:navigation-link {\\"label\\":\\"Fuga odio quis tempora\\",\\"type\\":\\"page\\",\\"description\\":\\"\\",\\"rel\\":\\"\\",\\"id\\":56,\\"url\\":\\"http://localhost:8889/?page_id=56\\",\\"title\\":\\"\\",\\"kind\\":\\"post-type\\",\\"className\\":\\"\\"} -->
+<!-- wp:navigation-link {\\"label\\":\\"In consectetur repellendus eveniet maiores aperiam\\",\\"type\\":\\"page\\",\\"description\\":\\"\\",\\"rel\\":\\"\\",\\"id\\":15,\\"url\\":\\"http://localhost:8889/?page_id=15\\",\\"title\\":\\"\\",\\"kind\\":\\"post-type\\",\\"className\\":\\"\\"} -->
+<!-- wp:navigation-link {\\"label\\":\\"Mollitia maiores consequatur ea dolorem blanditiis\\",\\"type\\":\\"page\\",\\"description\\":\\"\\",\\"rel\\":\\"\\",\\"id\\":45,\\"url\\":\\"http://localhost:8889/?page_id=45\\",\\"title\\":\\"\\",\\"kind\\":\\"post-type\\",\\"className\\":\\"\\"} -->
+<!-- wp:navigation-link {\\"label\\":\\"Necessitatibus nisi qui qui necessitatibus quaerat possimus\\",\\"type\\":\\"page\\",\\"description\\":\\"\\",\\"rel\\":\\"\\",\\"id\\":27,\\"url\\":\\"http://localhost:8889/?page_id=27\\",\\"title\\":\\"\\",\\"kind\\":\\"post-type\\",\\"className\\":\\"\\"} /-->
 <!-- /wp:navigation-link -->
 <!-- /wp:navigation-link -->
 <!-- /wp:navigation-link -->
 <!-- /wp:navigation-link -->
 
-<!-- wp:navigation-link {\\"label\\":\\"Nulla omnis autem dolores eligendi\\",\\"description\\":\\"\\",\\"rel\\":\\"\\",\\"url\\":\\"http://localhost:8889/?page_id=43\\",\\"title\\":\\"\\",\\"className\\":\\"\\"} /-->
+<!-- wp:navigation-link {\\"label\\":\\"Nulla omnis autem dolores eligendi\\",\\"type\\":\\"page\\",\\"description\\":\\"\\",\\"rel\\":\\"\\",\\"id\\":43,\\"url\\":\\"http://localhost:8889/?page_id=43\\",\\"title\\":\\"\\",\\"kind\\":\\"post-type\\",\\"className\\":\\"\\"} /-->
 
-<!-- wp:navigation-link {\\"label\\":\\"Sample Page\\",\\"description\\":\\"\\",\\"rel\\":\\"\\",\\"url\\":\\"http://localhost:8889/?page_id=2\\",\\"title\\":\\"\\",\\"className\\":\\"\\"} /-->
+<!-- wp:navigation-link {\\"label\\":\\"Sample Page\\",\\"type\\":\\"page\\",\\"description\\":\\"\\",\\"rel\\":\\"\\",\\"id\\":2,\\"url\\":\\"http://localhost:8889/?page_id=2\\",\\"title\\":\\"\\",\\"kind\\":\\"post-type\\",\\"className\\":\\"\\"} /-->
 
-<!-- wp:navigation-link {\\"label\\":\\"Beatae qui labore voluptas eveniet officia quia voluptas qui porro sequi et aut est\\",\\"description\\":\\"Ratione nemo ut aut ullam sed assumenda quis est exercitationem\\",\\"rel\\":\\"\\",\\"url\\":\\"http://localhost:8889/?cat=7\\",\\"title\\":\\"\\",\\"className\\":\\"\\"} -->
-<!-- wp:navigation-link {\\"label\\":\\"Et minus itaque velit tempore hic quisquam saepe quas asperiores\\",\\"description\\":\\"Vel fuga enim rerum perspiciatis sapiente mollitia magni ut molestiae labore quae quia quia libero perspiciatis voluptatem quidem deleniti eveniet laboriosam doloribus dolor laborum accusantium modi ducimus itaque rerum cum nostrum\\",\\"rel\\":\\"\\",\\"url\\":\\"http: ...
-<!-- wp:navigation-link {\\"label\\":\\"Et quas a et mollitia et voluptas optio voluptate quia quo unde aut in nostrum iste impedit quisquam id aut\\",\\"description\\":\\"Quas sit labore earum omnis eos sint iste est possimus harum aut soluta sint optio quos distinctio inventore voluptate non ut aliquam ad ut voluptates fugiat numquam magnam modi repellendus modi laudantium et debitis officia es ...
-<!-- wp:navigation-link {\\"label\\":\\"Illo quis sit impedit itaque expedita earum deserunt magni doloremque velit eum id error\\",\\"description\\":\\"Doloremque vero sunt officiis iste voluptatibus voluptas molestiae sint asperiores recusandae amet praesentium et explicabo nesciunt similique voluptatum laudantium amet officiis quas distinctio quis enim nihil tempora\\",\\"rel\\":\\"\\",\\"url\ ...
+<!-- wp:navigation-link {\\"label\\":\\"Beatae qui labore voluptas eveniet officia quia voluptas qui porro sequi et aut est\\",\\"type\\":\\"category\\",\\"description\\":\\"Ratione nemo ut aut ullam sed assumenda quis est exercitationem\\",\\"rel\\":\\"\\",\\"id\\":7,\\"url\\":\\"http://localhost:8889/?cat=7\\",\\"title\\":\\"\\",\\"kind\\":\\"taxonomy\\",\\"className\\":\\"\\"} -->
+<!-- wp:navigation-link {\\"label\\":\\"Et minus itaque velit tempore hic quisquam saepe quas asperiores\\",\\"type\\":\\"category\\",\\"description\\":\\"Vel fuga enim rerum perspiciatis sapiente mollitia magni ut molestiae labore quae quia quia libero perspiciatis voluptatem quidem deleniti eveniet laboriosam doloribus dolor laborum accusantium modi ducimus itaque rerum cum nostrum\\",\\"rel\\" ...
+<!-- wp:navigation-link {\\"label\\":\\"Et quas a et mollitia et voluptas optio voluptate quia quo unde aut in nostrum iste impedit quisquam id aut\\",\\"type\\":\\"category\\",\\"description\\":\\"Quas sit labore earum omnis eos sint iste est possimus harum aut soluta sint optio quos distinctio inventore voluptate non ut aliquam ad ut voluptates fugiat numquam magnam modi repellendus modi laudan ...
+<!-- wp:navigation-link {\\"label\\":\\"Illo quis sit impedit itaque expedita earum deserunt magni doloremque velit eum id error\\",\\"type\\":\\"category\\",\\"description\\":\\"Doloremque vero sunt officiis iste voluptatibus voluptas molestiae sint asperiores recusandae amet praesentium et explicabo nesciunt similique voluptatum laudantium amet officiis quas distinctio quis enim nihil tempora\\ ...
 <!-- /wp:navigation-link -->
 <!-- /wp:navigation-link -->
 <!-- /wp:navigation-link -->
 
-<!-- wp:navigation-link {\\"label\\":\\"WordPress.org\\",\\"description\\":\\"\\",\\"rel\\":\\"\\",\\"url\\":\\"https://wordpress.org\\",\\"title\\":\\"\\",\\"className\\":\\"\\"} -->
-<!-- wp:navigation-link {\\"label\\":\\"Google\\",\\"description\\":\\"\\",\\"rel\\":\\"\\",\\"url\\":\\"https://google.com\\",\\"title\\":\\"\\",\\"className\\":\\"\\"} /-->
+<!-- wp:navigation-link {\\"label\\":\\"WordPress.org\\",\\"type\\":\\"custom\\",\\"description\\":\\"\\",\\"rel\\":\\"\\",\\"id\\":108,\\"url\\":\\"https://wordpress.org\\",\\"title\\":\\"\\",\\"kind\\":\\"custom\\",\\"className\\":\\"\\"} -->
+<!-- wp:navigation-link {\\"label\\":\\"Google\\",\\"type\\":\\"custom\\",\\"description\\":\\"\\",\\"rel\\":\\"\\",\\"id\\":109,\\"url\\":\\"https://google.com\\",\\"title\\":\\"\\",\\"kind\\":\\"custom\\",\\"className\\":\\"\\"} /-->
 <!-- /wp:navigation-link -->
 <!-- /wp:navigation -->"
```
All changes due to setting of block attributes when converting from menus.

getdave commented on packages/e2e-tests/specs/experiments/blocks/__snapshots__/navigation.test.js.snap:
```diff
@@ -2,36 +2,36 @@
-<!-- wp:navigation-link {\\"label\\":\\"Fuga odio quis tempora\\",\\"url\\":\\"http://localhost:8889/?page_id=56\\"} -->
-<!-- wp:navigation-link {\\"label\\":\\"In consectetur repellendus eveniet maiores aperiam\\",\\"url\\":\\"http://localhost:8889/?page_id=15\\"} -->
-<!-- wp:navigation-link {\\"label\\":\\"Mollitia maiores consequatur ea dolorem blanditiis\\",\\"url\\":\\"http://localhost:8889/?page_id=45\\"} -->
-<!-- wp:navigation-link {\\"label\\":\\"Necessitatibus nisi qui qui necessitatibus quaerat possimus\\",\\"url\\":\\"http://localhost:8889/?page_id=27\\"} /-->
+<!-- wp:navigation-link {\\"label\\":\\"Est ea vero non nihil officiis in\\",\\"type\\":\\"page\\",\\"id\\":53,\\"url\\":\\"http://localhost:8889/?page_id=53\\",\\"kind\\":\\"post-type\\"} -->
+<!-- wp:navigation-link {\\"label\\":\\"Fuga odio quis tempora\\",\\"type\\":\\"page\\",\\"id\\":56,\\"url\\":\\"http://localhost:8889/?page_id=56\\",\\"kind\\":\\"post-type\\"} -->
+<!-- wp:navigation-link {\\"label\\":\\"In consectetur repellendus eveniet maiores aperiam\\",\\"type\\":\\"page\\",\\"id\\":15,\\"url\\":\\"http://localhost:8889/?page_id=15\\",\\"kind\\":\\"post-type\\"} -->
+<!-- wp:navigation-link {\\"label\\":\\"Mollitia maiores consequatur ea dolorem blanditiis\\",\\"type\\":\\"page\\",\\"id\\":45,\\"url\\":\\"http://localhost:8889/?page_id=45\\",\\"kind\\":\\"post-type\\"} -->
+<!-- wp:navigation-link {\\"label\\":\\"Necessitatibus nisi qui qui necessitatibus quaerat possimus\\",\\"type\\":\\"page\\",\\"id\\":27,\\"url\\":\\"http://localhost:8889/?page_id=27\\",\\"kind\\":\\"post-type\\"} /-->
 <!-- /wp:navigation-link -->
 <!-- /wp:navigation-link -->
 <!-- /wp:navigation-link -->
 <!-- /wp:navigation-link -->
 
-<!-- wp:navigation-link {\\"label\\":\\"Nulla omnis autem dolores eligendi\\",\\"url\\":\\"http://localhost:8889/?page_id=43\\"} /-->
+<!-- wp:navigation-link {\\"label\\":\\"Nulla omnis autem dolores eligendi\\",\\"type\\":\\"page\\",\\"id\\":43,\\"url\\":\\"http://localhost:8889/?page_id=43\\",\\"kind\\":\\"post-type\\"} /-->
 
-<!-- wp:navigation-link {\\"label\\":\\"Sample Page\\",\\"url\\":\\"http://localhost:8889/?page_id=2\\"} /-->
+<!-- wp:navigation-link {\\"label\\":\\"Sample Page\\",\\"type\\":\\"page\\",\\"id\\":2,\\"url\\":\\"http://localhost:8889/?page_id=2\\",\\"kind\\":\\"post-type\\"} /-->
 
-<!-- wp:navigation-link {\\"label\\":\\"Beatae qui labore voluptas eveniet officia quia voluptas qui porro sequi et aut est\\",\\"description\\":\\"Ratione nemo ut aut ullam sed assumenda quis est exercitationem\\",\\"url\\":\\"http://localhost:8889/?cat=7\\"} -->
-<!-- wp:navigation-link {\\"label\\":\\"Et minus itaque velit tempore hic quisquam saepe quas asperiores\\",\\"description\\":\\"Vel fuga enim rerum perspiciatis sapiente mollitia magni ut molestiae labore quae quia quia libero perspiciatis voluptatem quidem deleniti eveniet laboriosam doloribus dolor laborum accusantium modi ducimus itaque rerum cum nostrum\\",\\"url\\":\\"http://localhost:8889/ ...
-<!-- wp:navigation-link {\\"label\\":\\"Et quas a et mollitia et voluptas optio voluptate quia quo unde aut in nostrum iste impedit quisquam id aut\\",\\"description\\":\\"Quas sit labore earum omnis eos sint iste est possimus harum aut soluta sint optio quos distinctio inventore voluptate non ut aliquam ad ut voluptates fugiat numquam magnam modi repellendus modi laudantium et debitis officia es ...
-<!-- wp:navigation-link {\\"label\\":\\"Illo quis sit impedit itaque expedita earum deserunt magni doloremque velit eum id error\\",\\"description\\":\\"Doloremque vero sunt officiis iste voluptatibus voluptas molestiae sint asperiores recusandae amet praesentium et explicabo nesciunt similique voluptatum laudantium amet officiis quas distinctio quis enim nihil tempora\\",\\"url\\":\\"http://loca ...
+<!-- wp:navigation-link {\\"label\\":\\"Beatae qui labore voluptas eveniet officia quia voluptas qui porro sequi et aut est\\",\\"type\\":\\"category\\",\\"description\\":\\"Ratione nemo ut aut ullam sed assumenda quis est exercitationem\\",\\"id\\":7,\\"url\\":\\"http://localhost:8889/?cat=7\\",\\"kind\\":\\"taxonomy\\"} -->
+<!-- wp:navigation-link {\\"label\\":\\"Et minus itaque velit tempore hic quisquam saepe quas asperiores\\",\\"type\\":\\"category\\",\\"description\\":\\"Vel fuga enim rerum perspiciatis sapiente mollitia magni ut molestiae labore quae quia quia libero perspiciatis voluptatem quidem deleniti eveniet laboriosam doloribus dolor laborum accusantium modi ducimus itaque rerum cum nostrum\\",\\"id\\": ...
+<!-- wp:navigation-link {\\"label\\":\\"Et quas a et mollitia et voluptas optio voluptate quia quo unde aut in nostrum iste impedit quisquam id aut\\",\\"type\\":\\"category\\",\\"description\\":\\"Quas sit labore earum omnis eos sint iste est possimus harum aut soluta sint optio quos distinctio inventore voluptate non ut aliquam ad ut voluptates fugiat numquam magnam modi repellendus modi laudan ...
+<!-- wp:navigation-link {\\"label\\":\\"Illo quis sit impedit itaque expedita earum deserunt magni doloremque velit eum id error\\",\\"type\\":\\"category\\",\\"description\\":\\"Doloremque vero sunt officiis iste voluptatibus voluptas molestiae sint asperiores recusandae amet praesentium et explicabo nesciunt similique voluptatum laudantium amet officiis quas distinctio quis enim nihil tempora\\ ...
 <!-- /wp:navigation-link -->
 <!-- /wp:navigation-link -->
 <!-- /wp:navigation-link -->
 
-<!-- wp:navigation-link {\\"label\\":\\"WordPress.org\\",\\"url\\":\\"https://wordpress.org\\"} -->
-<!-- wp:navigation-link {\\"label\\":\\"Google\\",\\"url\\":\\"https://google.com\\"} /-->
+<!-- wp:navigation-link {\\"label\\":\\"WordPress.org\\",\\"type\\":\\"custom\\",\\"url\\":\\"https://wordpress.org\\",\\"kind\\":\\"custom\\"} -->
+<!-- wp:navigation-link {\\"label\\":\\"Google\\",\\"type\\":\\"custom\\",\\"url\\":\\"https://google.com\\",\\"kind\\":\\"custom\\"} /-->
 <!-- /wp:navigation-link -->
 <!-- /wp:navigation -->"
 `;
```
Changes due to setting block attributes such as `id` on entity block variations (eg: `Page`, `Category`) that need to store a reference to the database object which they represent.

getdave commented:
@gwwar I think this is now in a better place. Note the PR description which now has various caveats which were required to keep this PR size down. I'd really appreciate another review, particularly around Custom Post Types and how we can add some tests to cover these.

Mamaduka commented:
Hi, @getdave

I followed your testing instructions. I can see that original menu item types are correctly retained when saving the menu in the experimental editor. 🎉 

I see "Custom Link" variation for the following menu item types: "Tag," "Portfolio Item," "Portfolio Category," "Portfolio Tag." I expected those to be "Category Link" and "Post Link" based on your screencast.

getdave commented:
Thank you for the review 👍 

> I see "Custom Link" variation for the following menu item types: "Tag," "Portfolio Item," "Portfolio Category," "Portfolio Tag." I expected those to be "Category Link" and "Post Link" based on your screencast.

~Let me check those again. I might have broken something.~

@Mamaduka I checked and I found that saving in the Navigation Editor preserve the types for the custom `Portfolio Item/Category/Tag`.


https://user-images.githubusercontent.com/444434/116268911-a8cdbd80-a775-11eb-8ca7-c606cccf9498.mov

Did I miss something?

gwwar commented:
> Post type variations display as Post link in the block sidebar (screenshot). They should probably use the title of the Custom Post Type (eg: Portfolio link)

These are new post type labels `item_link` and `item_link_description` and should be added during CPT registration. If we glue together two translated strings, the translation is often very poor for specific languages. CPTs with missing labels fallback to the default post or category label.

gwwar commented:
We can probably cover pieces of this in a follow up PRs but wanted to note some extra options in the old menu editor (hiding under screen options):

<img width="971" alt="Screen Shot 2021-04-27 at 9 31 19 AM" src="https://user-images.githubusercontent.com/1270189/116278566-77d09700-a73b-11eb-9467-64b735ed5f3b.png">

Currently open in new tab doesn't appear to reflect between the two old menu nav/navigation editor:

| menu editor | navigation editor |
|-----|-----|
| <img width="459" alt="Screen Shot 2021-04-27 at 9 31 07 AM" src="https://user-images.githubusercontent.com/1270189/116278882-c2eaaa00-a73b-11eb-9319-3c3df6b6323b.png"> | <img width="516" alt="Screen Shot 2021-04-27 at 9 35 33 AM" src="https://user-images.githubusercontent.com/1270189/116279063-f3324880-a73b-11eb-975b-193463d59b66.png"> |

getdave commented:
> We can probably cover pieces of this in a follow up PRs but wanted to note some extra options in the old menu editor (hiding under screen options):
> 
> <img alt="Screen Shot 2021-04-27 at 9 31 19 AM" width="971" src="https://user-images.githubusercontent.com/1270189/116278566-77d09700-a73b-11eb-9467-64b735ed5f3b.png">
> 
> Currently open in new tab doesn't appear to reflect between the two old menu nav/navigation editor:
> 
> menu editor	navigation editor
> <img alt="Screen Shot 2021-04-27 at 9 31 07 AM" width="459" src="https://user-images.githubusercontent.com/1270189/116278882-c2eaaa00-a73b-11eb-9319-3c3df6b6323b.png">	<img alt="Screen Shot 2021-04-27 at 9 35 33 AM" width="516" src="https://user-images.githubusercontent.com/1270189/116279063-f3324880-a73b-11eb-975b-193463d59b66.png">

Agreed 👍 I will definitely be covering those in a separate PR to avoid this one getting too large. Thus PR adds the mechanics and fixes the immediate Issue. Follows will tie this all together.

gwwar commented on packages/block-library/src/navigation/placeholder.js:
```diff
@@ -56,6 +56,12 @@ function mapMenuItemsToBlocks( menuItems ) {
+				id: menuItem?.object_id,
+			} ),
+			// the equivalent on `nav_menu_item` is "type" but it is stored with underscores whereas the block attr uses hyphens.
+			kind: menuItem.type.replace( '_', '-' ),
```
🐍 snake vs kebab case. Are values here going to be `post-type`, `taxonomy` and now `custom`? Should the navigation block now also set the `custom` type?

Note that while the custom link variation is an option in the block editor, this really means that we don't set initial search parameters. Folks might end up adding a custom link or a specific type like a post link.

getdave replied:
The `kind` attribute on the block basically maps to [the `type` attribute on the nav menu item](https://core.trac.wordpress.org/browser/tags/5.7.1/src/wp-includes/nav-menu.php#L802). The main difference is that the values use `_` instead of `-` as separators.

Note that this is _not_ an identifier of the object type (ie: `Post`, `Page`, `Category`, `{Custom Post Type} Category`...etc) but rather an enum of:

* [`post-type`](https://core.trac.wordpress.org/browser/tags/5.7.1/src/wp-includes/nav-menu.php#L822).
* [`post-type-archive`](https://core.trac.wordpress.org/browser/tags/5.7.1/src/wp-includes/nav-menu.php#L862)
* [`taxonomy`](https://core.trac.wordpress.org/browser/tags/5.7.1/src/wp-includes/nav-menu.php#L877)

The use of `custom` is simply a convention rather than part of the `enum` which allows users to set a custom link without having to associate it with any given "entity" within WordPress. 

I would have thought it would be best for the block to use `custom` as the fallback case.

> Note that while the custom link variation is an option in the block editor, this really means that we don't set initial search parameters

This makes sense because you aren't searching for an entity.

> ...or a specific type like a post link.

I don't follow this sorry. Can you help me to understand?

gwwar commented on packages/edit-navigation/src/store/utils.js:
```diff
@@ -172,3 +187,85 @@ export function computeCustomizedAttribute(
+ * @param {string} blockAttributeValue the value of the block attribute to be mapped.
+ * @return {Object} block attribute mapped to correct menu item.
+ */
+export const mapBlockAttributeToMenuItemField = function mapBlockAttributeToMenuItemField(
```
Minor: What benefits do we get from using a named function expression here?

<img width="507" alt="Screen Shot 2021-04-27 at 1 44 52 PM" src="https://user-images.githubusercontent.com/1270189/116310459-e541ef00-a75e-11eb-8ded-53b2e53b61be.png">

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/name

gwwar commented on packages/edit-navigation/src/store/utils.js:
```diff
@@ -172,3 +187,85 @@ export function computeCustomizedAttribute(
+		// 4. `custom`.
+		kind: {
+			attr: 'type', // the equivalent on `nav_menu_item` is "type" but it is stored with underscores whereas the block attr uses hyphens.
+			mapper: () => blockAttributeValue.replace( '-', '_' ),
```
🤔 from the usage above, it might make sense to write this as an object to object transform vs per field. I think doing this per field adds some indirection and makes it a bit harder to read. A good exercise would be to port over the other items like title to see if what pattern we like better. Readability is subjective of course, so folks are welcome to chime in too

Note that data is also likely to be missing or malformed. A json schema might make sense in the transform function, though we'd prefer a decent attempt at fixing the data vs throwing an error (say if `kind` does not match the enum of values).

```jsx
const blockAttributesToMenuItem = ( { label, url, description, id, kind } ) => { 
    return {
        title: label,
        object_id: id,
        type: kind.replace( '-', '_' ),
        // and so on...
    }
}
```

getdave replied:
I've now changed this to use a model similar to that which you suggest here.

I went the other way mainly to avoid having to introduce too many changes into this PR but it ended up introducing unwanted complexity and indirection (as you correctly pointed out).

gwwar reviewed (commented):
This is looking good @getdave ✨  I still need to finish manual testing, but my overall take is that we can probably fix the most obvious things and set aside some other cases for immediate follow up.

I left a few suggestions on the transform patterns. We might want a few more 👀s there on what folks prefer for readability and handling of malformed data. (For example, I just spotted this one: https://github.com/WordPress/gutenberg/issues/31254 that I'll fix tomorrow unless someone gets to it first.)

Mamaduka commented:
Hi, @getdave

For some reason, I still see the same results. Attaching the screen recording:

https://user-images.githubusercontent.com/240569/116354498-943b0500-a809-11eb-8f71-49814fa0550c.mp4

Aljullu commented on packages/edit-navigation/src/store/utils.js:
```diff
@@ -172,3 +187,85 @@ export function computeCustomizedAttribute(
+};
+
+/**
+ * Converts a given key in an object to it's mapped equivalent.
```
Typo?
```suggestion
 * Converts a given key in an object to its mapped equivalent.
```

Aljullu reviewed (commented):
I did some testing of this PR with WooCommerce, because it has its own custom post types and taxonomies, and everything worked well on my end, I couldn't spot any issues. I also took a look at the code but will let the code review for somebody with more expertise.

gwwar commented:
Can't save a post-format in navigation editor:

This takes a bit of setup:

1. Activate a theme that supports post-formats, like Twenty Thirteen
2. Create some posts that use a post format like quote or gallery
3. Enable the navigation editor and create a new menu
4. Insert a post-format link (points to the post format archive). Note that this should be using the post-format variation:

https://user-images.githubusercontent.com/1270189/116439517-7e771100-a804-11eb-9a54-8ba904284db0.mp4

A custom link works out okay but we should avoid the 500 here:

| 500 | Sent values |
|-----|-----|
| <img width="1250" alt="Screen Shot 2021-04-28 at 9 30 31 AM" src="https://user-images.githubusercontent.com/1270189/116440165-0f4dec80-a805-11eb-84ae-442f20c1f928.png"> | <img width="842" alt="Screen Shot 2021-04-28 at 9 26 46 AM" src="https://user-images.githubusercontent.com/1270189/116440293-30164200-a805-11eb-91db-5c1f92a9929b.png"> |

gwwar commented:
Tag type is saved as "post_tag"

1. Visit the Navigation Editor
2. Add a tag variation and select a tag
3. Notice that this renders as a "custom" link type
4. Save and use the menu to prepopulate a navigation block in the post or site editor
5. type is saved as "post_tag" instead of "tag"

See general testing:

https://user-images.githubusercontent.com/1270189/116442513-6d7bcf00-a807-11eb-8d1c-37e290fae29a.mp4

gwwar commented:
Link Title is not persisted if we save a value in the navigation editor, then load in the navigation block on post/site editor

gwwar commented:
Pretty close @getdave! Thanks for tackling this one. I verified some CPTs, post, tags, category, pages, drafts created in the nav editor, and some custom formats like tel/mailto.

Some things I think we'll need before merge:
- [x] Get another code review on the per-field transform, I think it adds unnecessary indirection, but I'm okay to defer if other folks don't mind it. Eg it sticks out to me, but it's not blocking.
- [x] Fix [tag type serialization](https://github.com/WordPress/gutenberg/pull/31004#issuecomment-828613993)
- [x] Fix [link title serialization](https://github.com/WordPress/gutenberg/pull/31004#issuecomment-828634139)

Can be done as follow up or in this PR:
- [ ] Fix [post-format serialization](https://github.com/WordPress/gutenberg/pull/31004#issuecomment-828601727) (currently 500s)
- [x] Fix [open in new tab serialization](https://github.com/WordPress/gutenberg/pull/31004#issuecomment-827747951)

🤔 As a side note, I had trouble trying to setup a post type archive. I curious to see what this persisted as in the menu editor. 

Since we're trying to persist the custom type now, I also spun up https://github.com/WordPress/gutenberg/pull/31259 for the navigation block. 👀 are appreciated there.

For other folks reviewing if you're wondering where CPTs are in the menu editor, there's a lot of options hiding in "Screen Options"

<img width="1090" alt="Screen Shot 2021-04-28 at 10 13 53 AM" src="https://user-images.githubusercontent.com/1270189/116445273-8f2a8580-a80a-11eb-959a-f98b5738b107.png">

getdave committed:
Update packages/edit-navigation/src/store/utils.js

Co-authored-by: Albert Juhé Lluveras <contact@albertjuhe.com>

getdave commented:
> * Get another code review on the per-field transform, I think it adds unnecessary indirection, but I'm okay to defer if other folks don't mind it. Eg it sticks out to me, but it's not blocking.

I was using this approach in order to keep the PR limited in scope. However, in hindsight it has made things necessarily complex. I'm going to refactor along the lines you suggest and deal with all the attributes in one hit.

getdave committed:
Refactor to create per block/menuItem conversion rather than on per field/attr basis

getdave committed:
Doc blocks

getdave committed:
Reuse menu to blocks conversion util on navigation-link block

getdave committed:
Fix resolver unit tests

getdave committed:
Add new resolver test to assert on entity type conversion

getdave commented:
> * Fix [tag type serialization](https://github.com/WordPress/gutenberg/pull/31004#issuecomment-828613993)

I'm not sure we should look to fix in this PR. It's already getting quite large and fixing this would touch several other files. Moreover, the issue was not introduced by this PR and can be replicated in `trunk`. I have a fix and will create a separate PR in order that we can keep this one simple.

> Fix link title serialization
> Fix open in new tab serialization

There were a few issues here which should have been resolved as a result of the wider refactor.

getdave committed:
Update snapshots

getdave committed:
Fix nav editor snapshots to remove IDs from custom link types

getdave committed:
Avoid overwriting the ID attribute when calling onChange with no ID prop

getdave commented:
Ok @gwwar - thanks for your reviews and comments. Super helpful and much appreciated.

I've revised everything now and I think it's all working better.

The only things I'm not sure we need to cover in _this_ PR are:

* post formats
* post tags

These seem to both be existing issues and we'd be growing this PR very large if we tackle them here. 

Happy to raise and follow up immediately afterwards.

Will ping again in [the nav editor Slack channel ](https://wordpress.slack.com/archives/C01KDAZJMQ9) for review.

talldan commented on packages/edit-navigation/src/store/test/utils.js:
```diff
@@ -299,12 +341,269 @@ describe( 'computeCustomizedAttribute', () => {
+
+	const blockAttrs = blocks.map( ( block ) => block.attributes );
+
+	describe( 'mapping block attributes to menu item fields', () => {
```
Another option for these tests is to convert to A to B and then B back to A and compare the results. E.g. for block attributes:
```js
const blockAttributes = { // ... some fixture data };
const menuItem = blockAttributesToMenuItem( blockAttributes );
const blockAttributes2 = menuItemToBlockAttributes( menuItem );
expect( blockAttributes ).toEqual( blockAttributes2 );
```

And the same for menu items:
```js
const menuItem = { // ... some fixture data };
const blockAttributes = menuItemToBlockAttributes( menuItem );
const menuItem2 = blockAttributesToMenuItem( blockAttributes );
expect( menuItem ).toEqual( menuItem2 );
```

That will ensure no data is lost/corrupted in conversion. I would say it might be worth having that in addition to the other tests.

getdave replied:
Great idea @talldan. Thanks!

getdave replied:
I've decided to not include these types of tests because following [this discussion](https://github.com/WordPress/gutenberg/pull/31004#discussion_r623756348) we now no longer have lossless conversion in our code because we allow Core to set the defaults. Therefore a 1:1:1 conversion will no longer produce the results we expect.

talldan commented on packages/edit-navigation/src/store/utils.js:
```diff
@@ -172,3 +160,99 @@ export function computeCustomizedAttribute(
+	target,
+} ) => {
+	return {
+		label: menuItemTitleField.rendered,
+		rel: xfn.join( ' ' ).trim(),
+		className: classes.join( ' ' ).trim(),
+		title: attr_title,
+		type: object || 'custom',
+		...( 'custom' !== object && {
+			id: object_id,
+		} ),
+		description,
+		url,
+		kind: menuItemTypeField?.replace( '_', '-' ) ?? 'custom',
+		opensInNewTab: target === NEW_TAB_TARGET_ATTRIBUTE,
```
I wonder if the code should be a bit safer in handling some potentially `undefined` values here (`menuItemTitleField`, `xfn`, `classes`). I think I'd consider any menuItem data outside this code's control, so it should be ready for anything 😄 

Especially given we don't know what plugins or users have done previously to set these values.

Falling back to an empty string would probably be fine.

getdave replied:
Yes that is a very good point. I'll add some safety checks and fallbacks.

getdave replied:
See https://github.com/WordPress/gutenberg/pull/31004#discussion_r623756348

talldan commented on packages/edit-navigation/src/store/test/resolvers.js:
```diff
@@ -189,4 +231,105 @@ describe( 'getNavigationPostForMenu', () => {
+			},
+		];
+
+		// Feed a known value to the generator yield result
```
I always find the generator tests hard to read, so would be great for the comment to hint at what code this generator.next is for - e.g. which control/selector/action is being triggered.

getdave replied:
TBH I still find the [values being passed into `yield`](https://javascript.info/generators#yield-is-a-two-way-street) confusing as it can seem out of step. I've attempted to document the code to explain what is happening.

talldan commented on packages/block-library/src/navigation/map-menu-items-to-blocks.js:
```diff
@@ -0,0 +1,83 @@
+	target,
+} ) => {
+	return {
+		label: menuItemTitleField.rendered,
+		rel: xfn.join( ' ' ).trim(),
+		className: classes.join( ' ' ).trim(),
```
Similar comment here to the previous one, it'd be good to handle `undefined` values.

getdave replied:
I was looking at block attrs <--> menu item fields conversion. I couldn't decide whether we should:

1. Set a default value at the point of conversion.
2. Omit missing values on the source object on the target and allow them to fall through to whatever defaults are provided.

I've gone with option 2 because often the defaults are set by Core and thus we shouldn't set our own duplicate defaults here. 

For example block attrs -> menu item. [The menu item code in Core actually has a tonne of defaults](https://core.trac.wordpress.org/browser/tags/5.7.1/src/wp-includes/nav-menu.php#L438) so we don’t need to pass these. In fact there’s a good argument we should allow Core to define the defaults and just pass anything the block has overridden.

I've updated the code and associated tests to reflect this.

talldan reviewed (commented):
Looks really good so far, it's definitely nice to see this code move toward something more maintainable (and correct).

I did some testing by saving various types of menu items in the current menu screen, then loaded up the new nav screen, and then the reverse of that process.

There are a couple of things I've noticed:
- A tag created in the current menu screen becomes a custom Link when loaded in the new nav screen
- A label edited in the new nav screen isn't saved, loading up the current menu screen Custom Link labels are deleted and changes to other types aren't saved.

I didn't read all of the previous comments, so that may have been mentioned already.

getdave committed:
Adds full conversion cycle tests

getdave committed:
Fix invalid test that wasn’t performing correct assertion

getdave committed:
Refactor conversions to omit default values.

getdave committed:
Update tests to expect default or omitted values upon conversion

getdave committed:
Delete unused tests

getdave committed:
Code documentation for generator test steps

getdave committed:
Update snapshot to remove omitted attributes

getdave committed:
Update block’s mapping code and update associated e2e snapshot

getdave committed:
Fix bug with custom block label not being persisted correctly to nav menu item

getdave commented:
> A tag created in the current menu screen becomes a custom Link when loaded in the new nav screen

✅ Fixed

https://user-images.githubusercontent.com/444434/116694418-2d5b4e80-a9b7-11eb-922a-75146666fb62.mp4

Good catch. Another one which @gwwar also noticed. I've been through again and I was able to replicate. The issue is that the block variation is `tag` whereas WP Core uses `post_tag` when storing in the DB.

I've decided to avoid a block migration by simply doing the conversion on the fly when mapping blocks -> menu items (and vica versa). I've added x2 specific tests to provide coverage on these and documented the reasons behind the decision in code.

> * A label edited in the new nav screen isn't saved, loading up the current menu screen Custom Link labels are deleted and changes to other types aren't saved.

✅ Fixed

https://user-images.githubusercontent.com/444434/116694437-33e9c600-a9b7-11eb-956f-95013cd193c6.mov

Got it! I think @gwwar picked this one up as well. I've just worked out the cause, fixed the issue and added some test coverage.

getdave committed:
Fix mapping of “tag” to “post_tag” and add test coverage

getdave commented:
@gwwar @talldan I looked into the `post-format` issue. Looks like we have [a bug in the REST API response ](https://github.com/WordPress/gutenberg/issues/31380)which I can't fix in this PR.

Other than that I believe all the previous points have now been addressed, namely:

✅  Updated to use per block/menu item transforms rather than per property.
✅  Fix tag type serialization
✅  Fix link title serialization
✅  Fix open in new tab serialization

So the only issue left is `Fix post-format serialization (currently 500s)` which as I say above we can't address in this PR.

I'd say I'm ready for a final review 🤞 🙇

gwwar commented on packages/block-library/src/navigation/map-menu-items-to-blocks.js:
```diff
@@ -0,0 +1,97 @@
+ * @param {string} menuItem.description The description of this menu item.
+ * @param {string} menuItem.url The URL to which this menu item points.
+ * @param {string} menuItem.type The family of objects originally represented, such as 'post_type' or 'taxonomy'.
+ * @param {string} menuItem.target The target attribute of the link element for this menu item.
```
This is optional, but would it be easier to use a typescript `@typedef` definition for menuItem?

https://developer.wordpress.org/block-editor/contributors/code/coding-guidelines/#custom-types

getdave replied:
Done 👍

gwwar commented on packages/edit-navigation/src/store/utils.js:
```diff
@@ -172,3 +159,147 @@ export function computeCustomizedAttribute(
+ * @param {string} blockAttributes.rel the XFN relationship expressed in the link of this menu item.
+ * @param {string} blockAttributes.className the custom CSS classname attributes for this block.
+ * @param {string} blockAttributes.title the HTML title attribute for the block's link.
+ * @param {string} blockAttributes.type the type of variation of the block used (eg: 'Post', 'Custom', 'Category'...etc).
```
Also optional, since we can always follow up later, but a typedef also works well for any enums we need to define:

https://developer.wordpress.org/block-editor/contributors/code/coding-guidelines/#custom-types

getdave replied:
Created followup Issue.

getdave replied:
https://github.com/WordPress/gutenberg/issues/31450

gwwar commented on packages/edit-navigation/src/store/utils.js:
```diff
@@ -172,3 +159,147 @@ export function computeCustomizedAttribute(
+	return {
+		label: menuItemTitleField?.rendered || '',
+		type: object || 'custom',
+		kind: menuItemTypeField?.replace( '_', '-' ) || 'custom',
```
👏 thanks this is much easier to read in one place!

gwwar commented on packages/block-library/src/navigation-link/edit.js:
```diff
@@ -552,7 +552,12 @@ export default function NavigationLinkEdit( {
-										id,
+										// `id` represents the DB ID of the entity which this link represents (eg: Post ID).
+										// Therefore we must not inadvertently set it to `undefined` if the `onChange` is called with no `id` value.
+										// This is possible when a setting changes such as the `opensInNewTab`.
```
🤔 I think we'll want to unit test what's being passed through to on change in a different PR. It's likely that other fields may need a similar guard.

getdave replied:
Agreed. Indeed I raised this Issue for that

https://github.com/WordPress/gutenberg/issues/31379

gwwar commented on packages/edit-navigation/src/store/utils.js:
```diff
@@ -172,3 +159,147 @@ export function computeCustomizedAttribute(
+				object_id: id,
+			} ),
+		...( opensInNewTab && {
+			target: NEW_TAB_TARGET_ATTRIBUTE,
```
If opensInNewTab is false, target should be added as '' to allow the navigation editor to unset this value. This is because of how to how we're applying these values in https://github.com/WordPress/gutenberg/blob/7497fa6a524b4c366bea540b0dd619236dc89782/packages/edit-navigation/src/store/utils.js#L147-L149 

Menu editor ✅ 
Navigation Editor ❌ (Can't unset)
Post Editor ✅

I tested this in each editor with the following steps:

1. set opens in new tab
2. save
3. reload
4. verify opens in new tab
5. unset this value
6. save
7. reload




https://user-images.githubusercontent.com/1270189/116735837-eca60a00-a9a3-11eb-88f6-778bb2afaf89.mp4

getdave replied:
Great spot. Fixed and added specific test coverage.

gwwar reviewed (approved):
Nice work @getdave this is a huge improvement on our serialization and data handling in the navigation editor! ✨

I'm giving tentative approval. ✅ **Please fix the `opensInNewTab` case before merge**. No need for another look from me for that case if that's sorted.

I tested this by some manual smoke testing starting from various editors, and seeing if values were persisted starting from one editor or another.

Here's a quick example case:

https://user-images.githubusercontent.com/1270189/116737214-ced9a480-a9a5-11eb-84c8-19f340319cf2.mp4

gwwar commented on packages/block-library/src/navigation/map-menu-items-to-blocks.js:
```diff
@@ -0,0 +1,97 @@
+	return {
+		label: menuItemTitleField?.rendered || '',
+		type: object || 'custom',
+		kind: menuItemTypeField?.replace( '_', '-' ) || 'custom',
```
Looking at this again, maybe it makes more sense to only set `kind` to custom? Instead of both kind and type.

getdave replied:
Yep ok agreed. Done.

getdave committed:
Add typedefs for nav_menu_item

getdave committed:
Allow for unsetting of target attribute

getdave committed:
Remove “custom” default for block variation
